### ISSUE TITLE
Random noise in the discriminator

### ISSUE THREAD

**Issue by lbl680** [sefibk/KernelGAN#21]
Dear Author,
  I am confused about the random noise added into the input of the discriminator. Could you please explain the function of it?

**Comment by sefibk** [comment]
What is it you don't understand?
Why add noise or what kind of noise is added?

**Comment by lbl680** [comment]
In the 112th line of the KernelGAN.py

**Comment by sefibk** [comment]
Do you not understand the logic of adding noise or is it the function it self that you don't?

**Comment by lbl680** [comment]
the logic of adding noise

**Comment by sefibk** [comment]
D learns to differentiate between real images and generated ones and tries to learn which are real\fake.
Both types of images are 0-1 values as tensors, however, the real images have very specific values! 
Images are saved as integers 0-255, therefore, when represented as 0-1 tensors they will have only 256 specific values. Therefore, if no noise was added, D would learn these 256 values and differentiate the values very easily even if G would do a perfect job.
By adding noise to the real images, we discard this 256 specific values prior.
Did that answer your question?

**Comment by lbl680** [comment]
Get it. Thanks for your answer!

**Comment by nanmehta** [comment]
Thanks sir for the excellent paper, In regards to the answer for the logic of adding noise to the discriminator, you said that noise is being added to the real images.
But in the code ,noise is being added to the generated output.
d_pred_fake = self.D.forward((g_output + torch.randn_like(**g_output**) / 255.).detach())

**Comment by sefibk** [comment]
I guess I forgot this point. We added this noise just as good practice of GANs
